### PR TITLE
Move several jenkins legs to Windows only images

### PR DESCRIPTION
**Pull request by jaredpar**
This moves many of our Jenkins legs to be Windows only images (no Visual Studio).  The key part of enabling this change is setting up our repo to use an xcopy version of MSBuild.   In general our repo scripts will prefer machine installs of MSBuild but when unable to find one will use an xcopy version. 

Right now the MSBuild toolset is deployed via a zip file.  There are suggestions about using NuGet and I did look into that (see [this change](https://github.com/jaredpar/roslyn/commit/bec74a0f3e50251fe05c8dc23d2fbedf31c64048)).  Using a NuGet package is simple enough.  

The problem is that it's ~66MB in size due to the inclusion of framework reference assemblies.   In practice it didn't seem  NuGet was well setup to handle that size of a package.  Still open to using NuGet here.  I also considered breaking apart the reference assemblies and msbuild toolset into different packages but I'm not sure that will help much.

### CODE REVIEW

File: build/scripts/build-utils.ps1 [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+[string]$repoDir = [IO.Path]::GetFullPath($(join-path $PSScriptRoot "..\.."))
+[string]$binariesDir = join-path $repoDir "Binaries"
+[string]$scriptDir = $PSScriptRoot
+
+# The intent of this script is to locate and return the path to the MSBuild directory that
+# we should use for bulid operations.  The preference order for MSBuild to use is as 
+# follows
+#
+#   1. MSBuild from an active VS command prompt
+#   2. MSBuild from a machine wide VS install
```
Comment (tmat) [inline build/scripts/build-utils.ps1]: Why don't we always use the XCopy toolset?

File: build/scripts/deploy-msbuild.ps1 [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+$toolsDir = join-path $repoDir "Binaries\Toolset"
+$scratchDir = join-path $toolsDir "Scratch"
+
+try {
+
+    mkdir $toolsDir -ErrorAction silent | out-null
+    mkdir $scratchDir -ErrorAction silent | out-null
+    pushd $toolsDir
+
+    $scratchFile = join-path $scratchDir "msbuild.txt"
```
Comment (tmat) [inline build/scripts/deploy-msbuild.ps1]: >$scratchFile = join-path $scratchDir "msbuild.txt" [](start = 4, length = 50)

it'd make more sense to me if this was stored in Toolset\msbuild\version.txt

File: build/scripts/build-utils.ps1 (hunk 1 above)
Comment (jaredpar) [inline build/scripts/build-utils.ps1]: @tmat 

My long term goal is to have the opposite model: default is xcopy, let developers opt into using a machine wide install.

But this is a fairly new endeavor and I'm imagining it will take a week or two to work out all of the kinks.  Hence I want to start conservatively.

File: build/scripts/deploy-msbuild.ps1 (hunk 2 above)
Comment (jaredpar) [inline build/scripts/deploy-msbuild.ps1]: @tmat 

It made more sense to me as well and hence I started with that.  But then you quickly find out that our unzip APIs get really upset about files that exist in the directory.  😦

Comment (tmat) [inline build/scripts/deploy-msbuild.ps1]: You mean when updating to a newer version? I'd just delete the entire dir before unzipping.

Comment (jaredpar) [inline build/scripts/deploy-msbuild.ps1]: Hmm, true.

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-infrastructure for review

Note: the Code DOM change is being reviewed as a [separate PR](https://github.com/dotnet/roslyn/pull/17709).  I expect that commit to be rebased out later today.

File: build/scripts/build-utils.ps1 [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        return
+    }
+
+    # Look for a valid VS installation
+    try {
+        # It's important to execute locate-vs using a new powershell process.  This script will 
+        # use assemblies from our NuGet cache to locate VS.  This will cause them to be locked on
+        # disk for the duration of the powershell session.  Using a separate process makes sure
+        # they are not locked to this session which can be long lived.
+        $p = & powershell -noprofile -executionPolicy RemoteSigned -file (join-path $PSScriptRoot "locate-vs.ps1")
```
Comment (tmat) [inline build/scripts/build-utils.ps1]: Wouldn't it be simpler to add a CLI tool to LocateVS package?

Comment (jaredpar) [inline build/scripts/build-utils.ps1]: Yes much simpler.  If LocateVS had that tool I'd much prefer to use it.  

CC @tannergooding in case he wants to add one.

File: build/scripts/locate-vs.ps1 [hunk 4]
```diff
@@ -5,53 +5,43 @@ Param(
...
-  $requiredPackageIds += "Microsoft.Component.MSBuild"
-  $requiredPackageIds += "Microsoft.Net.Component.4.6.TargetingPack"
-  $requiredPackageIds += "Microsoft.VisualStudio.Component.PortableLibrary"
-  $requiredPackageIds += "Microsoft.VisualStudio.Component.Roslyn.Compiler"
-  $requiredPackageIds += "Microsoft.VisualStudio.Component.VSSDK"
+    $requiredPackageIds += "Microsoft.Component.MSBuild"
+    $requiredPackageIds += "Microsoft.Net.Component.4.6.TargetingPack"
+    $requiredPackageIds += "Microsoft.VisualStudio.Component.PortableLibrary"
+    $requiredPackageIds += "Microsoft.VisualStudio.Component.Roslyn.Compiler"
+    $requiredPackageIds += "Microsoft.VisualStudio.Component.VSSDK"
```
Comment (tmat) [inline build/scripts/locate-vs.ps1]: Seems like LocateVS API could use this set of package ids as the default value.

File: build/scripts/deploy-nuget.ps1 [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    $scratchFile = join-path $scratchDir "nuget.txt"
+    $scratchVersion = gc -raw $scratchFile -ErrorAction SilentlyContinue
+    if ($scratchVersion -eq ($version + [Environment]::NewLine)) { 
+        exit 0
+    }
+
+    $tempFileName = "nuget-$version.zip"
+    $tempFile = join-path $scratchDir $tempFileName
+    if (-not (test-path $tempFile)) { 
+        $url = "https://dist.nuget.org/win-x86-commandline/v$version/NuGet.exe"
```
Comment (tmat) [inline build/scripts/deploy-nuget.ps1]: Why do we need nuget.exe if we have ```msbuild /t:Restore```?

File: build/scripts/deploy-msbuild.ps1 [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    $scratchVersion = gc -raw $scratchFile -ErrorAction SilentlyContinue
+    if ($scratchVersion -eq ($version + [Environment]::NewLine)) { 
+        exit 0
+    }
+
+    $tempFileName = "msbuild-$version.zip"
+    $tempFile = join-path $scratchDir $tempFileName
+    if (-not (test-path $tempFile)) { 
+        $url = "https://jdashstorage.blob.core.windows.net/msbuild/msbuild-$version.zip"
```
Comment (tmat) [inline build/scripts/deploy-msbuild.ps1]: The zip file has unnecessary "msbuild" root dir.

Comment (tmat) [inline build/scripts/deploy-msbuild.ps1]: dotnet SDKs are missing from the xcopy msbuild.

Comment (jaredpar) [inline build/scripts/deploy-msbuild.ps1]: That's deliberate at this point.  I didn't want to add anything until we needed it.  Otherwise it's pretty much impossible to determine if I added it correctly.

File: build/scripts/deploy-nuget.ps1 (hunk 5 above)
Comment (jaredpar) [inline build/scripts/deploy-nuget.ps1]: We still have a number of plain old project.json files that we restore.

**Comment by tannergooding** [comment]
test windows_debug_vs-integration_prtest please

**Comment by tannergooding** [comment]
test windows_release_vs-integration_prtest please

File: build/scripts/build-utils.ps1 [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        write-output (locate-msbuild-xcopy)
+        return
+    }
+
+    # MSBuild from an active VS command prompt.  
+    if (${env:VSINSTALLDIR} -ne $null) {
+
+        # This line deliberately avoids using -ErrorAction.  Inside a VS command prompt
+        # an MSBuild command should always be available.
+        $p = (get-command msbuild).Source
```
Comment (tannergooding) [inline build/scripts/build-utils.ps1]: I think this will fail if the user has two versions of MSBuild on the path.

File: build/scripts/build-utils.ps1 (hunk 3 above)
Comment (tannergooding) [inline build/scripts/build-utils.ps1]: There is now an 'official' tool (that is one I didn't write and that is in the Microsoft org, rather than a Roslyn specific repo): https://github.com/Microsoft/vswhere

File: build/scripts/build-utils.ps1 [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+    # Look for a valid VS installation
+    try {
+        # It's important to execute locate-vs using a new powershell process.  This script will 
+        # use assemblies from our NuGet cache to locate VS.  This will cause them to be locked on
+        # disk for the duration of the powershell session.  Using a separate process makes sure
+        # they are not locked to this session which can be long lived.
+        $p = & powershell -noprofile -executionPolicy RemoteSigned -file (join-path $PSScriptRoot "locate-vs.ps1")
+        if ($? -and (test-path $p)) {
+            $p = join-path $p "..\..\MSBuild\15.0\Bin"
```
Comment (tannergooding) [inline build/scripts/build-utils.ps1]: Is this right? I thought I passed back VSINSTALLDIR rather than VSCOMNTOOLS?

File: build/scripts/build-utils.ps1 [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+            $p = [IO.Path]::GetFullPath($p)
```
Comment (tannergooding) [inline build/scripts/build-utils.ps1]: nit: Use `Resolve-Path $p` or `Resolve-Path -Path $p`

File: build/scripts/build-utils.ps1 [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        # machine.
+    }
+
+    write-output "xcopy"
+    write-output (locate-msbuild-xcopy)
+    return
+}
+
+function locate-msbuild() {
+    param ([switch]$xcopy = $false)
```
Comment (tannergooding) [inline build/scripts/build-utils.ps1]: why did you define it this way instead of `function locate-msbuild([[switch]$xcopy = $false)`?

File: build/scripts/build-utils.ps1 [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+    write-output "xcopy"
+    write-output (locate-msbuild-xcopy)
+    return
+}
+
+function locate-msbuild() {
+    param ([switch]$xcopy = $false) 
+    $both = locate-msbuild-core -xcopy:$xcopy
+    return $both[1]
```
Comment (tannergooding) [inline build/scripts/build-utils.ps1]: Why are you returning an array here?

File: build/scripts/build-utils.ps1 [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+# Locate the xcopy version of MSBuild
+function locate-msbuild-xcopy() {
+    # Use the xcopy toolset
+    & (join-path $PSScriptRoot "deploy-msbuild.ps1")
+    $p = join-path $PSScriptRoot "..\..\Binaries\Toolset\msbuild"
+    $p = [IO.Path]::GetFullPath($p)
```
Comment (tannergooding) [inline build/scripts/build-utils.ps1]: nit: Use Resolve-Path

File: build/scripts/build-utils.ps1 (hunk 3 above)
Comment (tmat) [inline build/scripts/build-utils.ps1]: Yes, we should definitely use vswhere

File: build/scripts/build-utils.ps1 [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+            ${env:TargetFrameworkRootPath}=(locate-framework-xcopy)
+            break;
+        }
+        "vscmd" {
+            # Nothing to do here as the VS Dev CMD should set all appropriate environment
+            # variables.
+            break;
+        }
+        "vsinstall" {
+            ${env:PATH}="$msbuildDir;${env:PATH}"
```
Comment (tannergooding) [inline build/scripts/build-utils.ps1]: Why this instead of launching Developer Command Prompt, which should give us the correct state?

File: build/scripts/build-utils.ps1 (hunk 3 above)
Comment (tmat) [inline build/scripts/build-utils.ps1]: I suggest we download the tool from ```https://github.com/Microsoft/vswhere/releases/download/1.0.47/vswhere.exe```
and run it to discover msbuild.

File: build/scripts/build-utils.ps1 (hunk 11 above)
Comment (tannergooding) [inline build/scripts/build-utils.ps1]: nevermind, I missed the comment on returning two values.

File: build/scripts/build-utils.ps1 (hunk 3 above)
Comment (jaredpar) [inline build/scripts/build-utils.ps1]: @tmat @tannergooding 

Seems reasonable.  I will look at doing that.

File: build/scripts/deploy-msbuild.ps1 [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+# This script will ensure that the MSBuild toolset of the specified 
+# version is downloaded and deployed to Binaries\Toolset\MSBuild
+#
+# This is a no-op if it is already present at the appropriate version
+
+param ()
+set-strictmode -version 2.0
+$ErrorActionPreference="Stop"
+
+$version = "0.1.7-alpha"
```
Comment (tannergooding) [inline build/scripts/deploy-msbuild.ps1]: nit: Make version a parameter with a default value of "0.1.7-alpha", it makes it easier to test new changes

File: build/scripts/build-utils.ps1 (hunk 3 above)
Comment (jaredpar) [inline build/scripts/build-utils.ps1]: There is a NuGet package as well.  Going to use that. 

https://www.nuget.org/packages/vswhere

File: build/scripts/build-utils.ps1 (hunk 7 above)
Comment (jaredpar) [inline build/scripts/build-utils.ps1]: I'm fine with that for now.  If it becomes an issue we can look into it.

File: build/scripts/deploy-msbuild.ps1 [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    if ($scratchVersion -eq ($version + [Environment]::NewLine)) { 
+        exit 0
+    }
+
+    $tempFileName = "msbuild-$version.zip"
+    $tempFile = join-path $scratchDir $tempFileName
+    if (-not (test-path $tempFile)) { 
+        $url = "https://jdashstorage.blob.core.windows.net/msbuild/msbuild-$version.zip" 
+        write-host "Downloading $tempFileName"
+        $webClient = New-Object -TypeName "System.Net.WebClient"
```
Comment (tannergooding) [inline build/scripts/deploy-msbuild.ps1]: `Invoke-WebRequest`?

File: build/scripts/build-utils.ps1 (hunk 8 above)
Comment (jaredpar) [inline build/scripts/build-utils.ps1]: This is correct.  I've validated it on a number of machines.

File: build/scripts/deploy-msbuild.ps1 [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        $webClient.DownloadFile($url, $tempFile)
+    }
+    
+    write-host "Unpacking msbuild.zip"
+    rm -re (join-path $toolsDir "msbuild") -ErrorAction SilentlyContinue
```
Comment (tannergooding) [inline build/scripts/deploy-msbuild.ps1]: Depending on aliases is bad, use `Remote-Item` instead

File: build/scripts/deploy-msbuild.ps1 [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    Add-Type -AssemblyName System.IO.Compression.FileSystem
+    [IO.Compression.ZipFile]::ExtractToDirectory($tempFile, $toolsDir)
```
Comment (tannergooding) [inline build/scripts/deploy-msbuild.ps1]: `Expand-Archive`?

File: build/scripts/deploy-nuget.ps1 [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+$ErrorActionPreference="Stop"
+
+$version = "3.6.0-beta1"
+$repoDir = [IO.Path]::GetFullPath((join-path $PSScriptroot "..\.."))
+$toolsDir = join-path $repoDir "Binaries\Toolset"
+$scratchDir = join-path $toolsDir "Scratch"
+
+try {
+
+    mkdir $toolsDir -ErrorAction silent | out-null
```
Comment (tannergooding) [inline build/scripts/deploy-nuget.ps1]: Aliases are bad, use `New-Item`

File: build/scripts/build-utils.ps1 (hunk 10 above)
Comment (jaredpar) [inline build/scripts/build-utils.ps1]: This is the only way you could do it in Powershell V1 hence burned into my brain. 

I can't find a good reference for when they allowed the new syntax hence I haven't switched to it.

File: build/scripts/build-utils.ps1 (hunk 13 above)
Comment (jaredpar) [inline build/scripts/build-utils.ps1]: Because launching the developer command prompt does a lot more than just provide MSBuild support, it puts a huge number of other tools / settings into the environment.  That potentially, and in the past concretely, changed the output of the build and tests.  Using MSBuild only, because that's all we actually need, explicitly hepls make the tests more reliable.

File: cibuild.cmd [hunk 19]
```diff
@@ -63,6 +60,17 @@ if defined testBuildCorrectness (
     exit /b 0
 )
 
+REM Restore the solution and ensure MSBuild is available on the path for use in the remainder
+REM of the script.
+if "%BuildXCopy%" == "true" (
+    call "%RoslynRoot%build\scripts\EnsureMSBuild.cmd" -xcopy || goto :BuildFailed
+    call "%RoslynRoot%Restore.cmd" /msbuild "%RoslynRoot%Binaries\Toolset\msbuild" || goto :BuildFailed
```
Comment (jasonmalinowski) [inline cibuild.cmd]: Doesn't restore.cmd already discover it in the xcopy case?

File: build/Targets/Settings.props [hunk 20]
```diff
@@ -36,7 +36,7 @@
          $(NuGetPackageRoot)\Microsoft.Net.Compilers\$(ToolsetCompilerPackageVersion)\build\Microsoft.Net.Compilers.props -->
     <ToolsetCompilerPropsFilePath>$(MSBuildThisFileDirectory)Microsoft.Net.Compilers.props</ToolsetCompilerPropsFilePath>
 
-    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">14.0</VisualStudioVersion>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">15.0</VisualStudioVersion>
```
Comment (tannergooding) [inline build/Targets/Settings.props]: Because MicroBuild is still on 14.0

**Comment by jaredpar** [comment]
@jasonmalinowski 

> The only specific issue I see is the commenting out of GitLink with no explained reason why. Please add a comment there w/ a tracking bug or fix that up before merging this. 

Thanks, forgot about this.  Here is the conversation I had with the GitLink folks about handling an xcopy MSBuild environment.  

https://github.com/GitTools/GitLink/issues/145

I'm going to implement their suggestion as a separate change that is merged before this one.  

> The dependencies of what scripts call what, what assumes what, and what passes data to what feels very adhoc and difficult to follow

Agree.  The problem here is that we are basically in a transition from having our repo largely run by CMD scripts to being run by Powershell scripts.  The hard to follow nature is in large part passing data between these environments.  

I've been talking with @tannergooding about this and plan to transition us mostly to Powershell with very thin CMD wrappers.  Didn't want to do that in this change because I felt like it was conflatnig the issues.  It should be simpler after.  

> A high level "here's how the build is supposed to work" doc? 

Did you review the markdown file I added to this change?  That was meant to be a discussion on how our build worked now.  If it's lacking I can update that.

**Comment by jasonmalinowski** [comment]
@jaredpar: you meant the MSBuild markdown file? That did help (thanks for adding it!), but it's mostly the data-flowing-between-scripts problem that bugs me. If I was asked to fix a bug in here, I'd first spend an hour listing all the scripts and drawing arrows between all of them tracking the data flow. I think my ask is just prioritize that cleanup you're thinking about once this is in, with a focus on commenting and a "imagine if this were production code with public APIs" point of view.

(I think my other problem is I just hate powershell. But I know you'd be first in hating it too...)

**Comment by jaredpar** [comment]
> That did help (thanks for adding it!), but it's mostly the data-flowing-between-scripts problem that bugs me.

Yeah that's the hard part.  As I noted though I'm hoping that mostly goes away once we move to more Powershell.  May try and do that before I make this change so it lessens the burden here.

**Comment by tamlin-mike** [comment]
Random ideas: robocopy (it's now in the OS) instead of xcopy? Makefiles (nmake) for dependencies? At least plain Makefile's is something everyone can both understand and edit. May not be the fastest for a gazillion-file build system, but it solved the DAG dependency problem decades ago.

**Comment by jaredpar** [comment]
retest ubuntu_16_debug_prtest please

File: build/scripts/deploy-msbuild.ps1 [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+    $scratchFile = Join-Path $scratchDir "msbuild.txt"
+    $scratchVersion = Get-Content -raw $scratchFile -ErrorAction SilentlyContinue
+    if ($scratchVersion -eq ($version + [Environment]::NewLine)) { 
+        exit 0
+    }
+
+    $tempFileName = "msbuild-$version.zip"
+    $tempFile = Join-Path $scratchDir $tempFileName
+    if (-not (Test-Path $tempFile)) { 
+        $url = "https://jdashstorage.blob.core.windows.net/msbuild/msbuild-$version.zip"
```
Comment (jasonmalinowski) [inline build/scripts/deploy-msbuild.ps1]: Which Azure subscription is this blob store under? If it's not a Microsoft one can we move the package?

(feel free to take this offline if you can't share details here)

Comment (jaredpar) [inline build/scripts/deploy-msbuild.ps1]: It's my private azure subscription.  Was using it as a test bed for this change.  Now that I'm close to approval I broke this up into two NuGet packages so it will come through normal channels.

**Comment by jaredpar** [comment]
retest microbuild_prtest please

**Comment by jaredpar** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by jaredpar** [comment]
retest windows_debug_vs-integration_prtest please
